[docs] Update Storage Documentation (#1392)
qdrant/landing_page: qdrant-landing/content/documentation/concepts/storage.md

# Storage

> In a Qdrant collection, all data is divided into **segments**. Each segment contains several components: **vector** and **payload storages** as well as **vector** and **payload indexes**. Behind the scenes, each segment has an ID mapper that links internal and external IDs. 

Reviewer: Also here, there's a layer in between.

Collections have shards. There may be multiple copies of shards, which we call replicas. Shards have segments.

We normally mostly abstract this concept away, and segments aren't visible on the logical data level.

I understand we might want to describe segments somewhere, but describing it in the [optimizer](https://qdrant.tech/documentation/concepts/optimizer/) section may be better. They're closely related to the optimization process.

---

```suggestion
In a Qdrant shard, all data is divided into **segments**. Each segment contains several components: **vector** and **payload storages** as well as **vector** and **payload indexes**. Behind the scenes, each segment has an ID mapper that links internal and external IDs. 
```

# Storage

A Qdrant collection can be split into multiple shards. Each shard is a separate storage unit. By default, a collection is created with a single [shard](/documentation/guides/distributed_deployment/). They are used to distribute data across nodes in a cluster, enabling parallel processing and improving performance. 

From a storage perspective, data within a shard is organized into segments. Each segment independently manages its vector and payload storage, along with their respective indexes.

![Storage Architecture](/documentation/concepts/storage/storage-architecture.png)

Typically, data in segments does not overlap. If the same data point is stored in multiple segments, it won't cause issues because the search process includes a deduplication mechanism.

Segments are made up of:

> - Vector and Payload Storage

Reviewer: Also quantized vectors (if configured)

- Vector and Payload [Indexes](/documentation/concepts/indexing/)
- An ID mapper that links internal and external IDs

Segments can be either `appendable` or `non-appendable`:
- **Appendable segments**: You can add, delete, and query data freely.

> - **Non-appendable segments**: You can only read and delete data.

Reviewer: Payload data can also be updated in non-appendable segment. But maybe it's better to not mention that here.

> Each collection can have segments that are configured differently, but it must include at least one `appendable` segment.

Reviewer: ```suggestion
Each collection can have segments that are configured differently, but each replica has at least one `appendable` segment to support incoming writes.
```

## Vector Storage

Qdrant offers two main options for storing vectors, depending on your application's needs. The choice affects search speed and RAM usage:

- **In-Memory storage**: 
  - Stores all vectors in RAM.

>   - Offers the highest speed because disk access is only needed for saving data.

Reviewer: ```suggestion
  - Offers the highest speed because disk access is only needed for loading data on start, or for saving data if changed.
```

})
```

When you enable **On-Disk** storage, all vectors in a collection are stored on disk immediately. This is recommended if your Qdrant instance uses fast disks and handles large collections.

To configure when segments switch to **On-Disk** storage, use the `memmap_threshold` option. You can set this threshold in two ways:

1. **Globally**: Adjust the `memmap_threshold` parameter in the [configuration file](/documentation/guides/configuration/).

> 2. **Per Collection**: Set the threshold during the [creation](/documentation/concepts/collections/#create-collection) or [update](/documentation/concepts/collections/#update-collection-parameters) of each collection.

Reviewer: Let's drop this, since the plan is to deprecate this parameter anyway.

Instead we can say to explicitly set `on_disk: true` for:
- vectors (vector storage)
- hnsw
- payload (`on_disk_payload: true`)
- payload indices

})
```

The rule of thumb for setting the `memmap_threshold` is straightforward:

- **Balanced Use**: Set `memmap_threshold` equal to `indexing_threshold` (default is 20000). This way, the optimizer handles all thresholds together without extra runs.
- **High Write Load & Low RAM**: Set `memmap_threshold` lower than `indexing_threshold`, e.g., 10000. This prioritizes converting segments to Memmap storage before indexing.

> Additionally, **On-Disk** storage can be used for the HNSW index. To enable this, set the `hnsw_config.on_disk` parameter to `true` during collection [creation](/documentation/concepts/collections/#create-a-collection) or [updating](/documentation/concepts/collections/#update-collection-parameters).

Reviewer: Same as the above, let's drop `memmap_threshold` all together and focus on `on_disk`.

})
```

## Payload Storage

Qdrant supports two types of payload storage: **In-Memory** and **On-Disk**.

- **In-Memory Storage**:
  - Loads payload data into RAM at startup.
  - Uses disk and [RocksDB](https://rocksdb.org/) for persistence.
  - Fast access but requires a lot of RAM, especially for large payloads like text or images.

- **On-Disk Storage**:
  - Reads and writes payloads directly to [Gridstore](/articles/gridstore-key-value-storage/).
  - Requires less RAM but has higher access latency.
  - If querying vectors with payload-based conditions, create a payload index for each field to avoid disk access. Indexed fields are kept in RAM.

> You can choose the type of payload storage in the [configuration file](/documentation/guides/configuration/) or by setting the `on_disk_payload` parameter when [creating](/documentation/concepts/collections/#create-collection) a collection.

Reviewer: Let's put on-disk on top, since it's the default now.

## Versioning

Qdrant ensures data integrity through a two-stage process:

1. **Write-Ahead Log (WAL)**:
   - All data changes are first written to the WAL.
   - The WAL orders operations and assigns them a sequential number.
   - Changes in the WAL are safe from power loss.

2. **Segment Updates**:
   - Changes are then applied to segments.
   - Each segment keeps the latest version of changes and the version of each point.

>    - If a new change has a lower sequential number than the current version, it is ignored.

Reviewer: ```suggestion
   - If a new change has a lower (older) sequential number than the current version, it is ignored.
```

> This process allows Qdrant to restore storage safely from the WAL in case of an unexpected shutdown.

Reviewer: ```suggestion
This process allows Qdrant to restore storage safely from the WAL in case of an unexpected shutdown, by replaying everything from the WAL.
```
